File: node/src/notification.ts
// ============ Notification Constants ============

export type DeliveryMode = 'carrier' | 'self-delivery' | 'self-pickup';
export type ChannelType = 'email' | 'copy_url' | 'sms' | 'whatsapp';

export interface NotificationResult {
    channel: ChannelType;
    status: 'sent' | 'skipped' | 'failed' | 'url_only';
    message?: string;
    url?: string;
    error?: string;
}

export const DELIVERY_MODES: Array<{
    key: DeliveryMode;
    label: string;
    description: string;
}> = [
    { key: 'carrier', label: 'Carrier', description: 'Third-party carrier delivery (default)' },
    { key: 'self-delivery', label: 'Self-delivery', description: 'Merchant staff delivers' },
    { key: 'self-pickup', label: 'Self-pickup', description: 'Customer picks up at store/locker' },
];

export const NOTIFICATION_CHANNELS: Array<{
    key: ChannelType;
    label: string;
    icon: string;
    description: string;
    requires: 'email' | 'phone' | 'none';
}> = [
    { key: 'email', label: 'Email', icon: '\u{1F4E7}', description: 'Send signing invitation email (recommended)', requires: 'email' },
    { key: 'copy_url', label: 'Copy URL', icon: '\u{1F517}', description: 'Generate signing URL for manual sharing', requires: 'none' },
    { key: 'sms', label: 'SMS', icon: '\u{1F4F1}', description: 'Requires activated SMS provider', requires: 'phone' },
    { key: 'whatsapp', label: 'WhatsApp', icon: '\u{1F4AC}', description: 'Requires activated WhatsApp provider', requires: 'phone' },
];

export function validateChannelRequirements(
    channels: ChannelType[],
    recipient: { email?: string; phone?: string }
): string[] {
    const missing: string[] = [];
    const needEmail = channels.includes('email');
    const needPhone = channels.some((c) => c === 'sms' || c === 'whatsapp');
    if (needEmail && !recipient.email) missing.push('email');
    if (needPhone && !recipient.phone) missing.push('phone');
    return missing;
}

File: rust/src/notification.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub enum DeliveryMode {
    #[serde(rename = "carrier")]
    Carrier,
    #[serde(rename = "self-delivery")]
    SelfDelivery,
    #[serde(rename = "self-pickup")]
    SelfPickup,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub enum ChannelType {
    #[serde(rename = "email")]
    Email,
    #[serde(rename = "copy_url")]
    CopyUrl,
    #[serde(rename = "sms")]
    Sms,
    #[serde(rename = "whatsapp")]
    WhatsApp,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct NotificationResult {
    pub channel: ChannelType,
    pub status: String,
    pub message: Option<String>,
    pub url: Option<String>,
    pub error: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DeliveryModeItem {
    pub key: DeliveryMode,
    pub label: &'static str,
    pub description: &'static str,
}

#[derive(Debug, Clone, Serialize)]
pub struct NotificationChannelItem {
    pub key: ChannelType,
    pub label: &'static str,
    pub icon: &'static str,
    pub description: &'static str,
    pub requires: &'static [&'static str],
}

pub const DELIVERY_MODES: &[DeliveryModeItem] = &[
    DeliveryModeItem {
        key: DeliveryMode::Carrier,
        label: "Carrier",
        description: "Ship with a third-party carrier",
    },
    DeliveryModeItem {
        key: DeliveryMode::SelfDelivery,
        label: "Self Delivery",
        description: "Deliver using your own fleet",
    },
    DeliveryModeItem {
        key: DeliveryMode::SelfPickup,
        label: "Self Pickup",
        description: "Customer picks up at a pickup point",
    },
];

const EMAIL_REQUIRES: &[&str] = &["email"];
const PHONE_REQUIRES: &[&str] = &["phone"];
const NO_REQUIRES: &[&str] = &[];

pub const NOTIFICATION_CHANNELS: &[NotificationChannelItem] = &[
    NotificationChannelItem {
        key: ChannelType::Email,
        label: "Email",
        icon: "mail",
        description: "Send notification via email",
        requires: EMAIL_REQUIRES,
    },
    NotificationChannelItem {
        key: ChannelType::CopyUrl,
        label: "Copy URL",
        icon: "link",
        description: "Generate a shareable URL",
        requires: NO_REQUIRES,
    },
    NotificationChannelItem {
        key: ChannelType::Sms,
        label: "SMS",
        icon: "message-square",
        description: "Send notification via SMS",
        requires: PHONE_REQUIRES,
    },
    NotificationChannelItem {
        key: ChannelType::WhatsApp,
        label: "WhatsApp",
        icon: "phone",
        description: "Send notification via WhatsApp",
        requires: PHONE_REQUIRES,
    },
];

pub fn validate_channel_requirements(
    channels: &[ChannelType],
    recipient_email: Option<&str>,
    recipient_phone: Option<&str>,
) -> Vec<String> {
    let mut missing = Vec::new();
    for ch in channels {
        for item in NOTIFICATION_CHANNELS {
            if item.key == *ch {
                for &req_field in item.requires {
                    match req_field {
                        "email" if recipient_email.is_none() || recipient_email.unwrap().is_empty() => {
                            missing.push("email".to_string());
                        }
                        "phone" if recipient_phone.is_none() || recipient_phone.unwrap().is_empty() => {
                            missing.push("phone".to_string());
                        }
                        _ => {}
                    }
                }
            }
        }
    }
    missing.sort();
    missing.dedup();
    missing
}
